modyn/trainer_server/internal/dataset/local_dataset_handler.py

```diff
@@ -0,0 +1,121 @@
+        if self.maximum_keys_in_memory is not None:
+            # tuples are progressively accumulated in this ndarray and then dumped to file when the
+            # desired file size is reached
+            self.output_samples_list = np.empty(self.maximum_keys_in_memory, dtype=np.dtype("i8,f8"))
+
+    def inform_samples(self, input_samples_list: np.ndarray) -> None:
+        assert self.maximum_keys_in_memory is not None
+        assert self.output_samples_list is not None
+        for element in input_samples_list:
+            self.output_samples_list[self.current_sample_index] = element
+            self.current_sample_index += 1
+
+            if self.current_sample_index == self.maximum_keys_in_memory:
+                self._samples_ready()
+
+    def store_last_samples(self) -> None:
+        assert self.maximum_keys_in_memory is not None
+        if self.current_sample_index > 0:
+            self.output_samples_list = self.output_samples_list[: self.current_sample_index]  # remove empty elements
+            self._samples_ready()
+
+    def _samples_ready(self) -> None:
+        assert self.maximum_keys_in_memory is not None
+        assert self.output_samples_list is not None
+
+        number_worker_samples = self.current_sample_index // self.number_of_workers
+
+        for worker in range(self.number_of_workers):
+            worker_samples = (
+                self.output_samples_list[worker * number_worker_samples : (worker + 1) * number_worker_samples]
+                if worker != self.number_of_workers - 1
+                else self.output_samples_list[worker * number_worker_samples :]
+            )
+            path = self._get_file_name(
+                self.pipeline_id, self.trigger_id, partition_id=self.current_file_index, worker_id=worker
+            )
+            self._write_file(path, worker_samples)
+
+        self.current_sample_index = 0
+        self.output_samples_list = np.empty(self.maximum_keys_in_memory, dtype=np.dtype("i8,f8"))
+        self.current_file_index += 1
+
+    def get_keys_and_weights(
+        self,
+        partition_id: int,
+        worker_id: int,  # pylint: disable=unused-argument
+    ) -> tuple[list[int], list[float]]:
+        path = self._get_file_name(self.pipeline_id, self.trigger_id, partition_id, worker_id)
+        file = path.parent / (path.name + ".npy")
+        tuples_list = self._parse_file(file)
+
+        keys = []
+        weights = []
+        for key, weight in tuples_list:
+            keys.append(key)
+            weights.append(weight)
+        return keys, weights
+
+    def clean_working_directory(self) -> None:
+        if os.path.isdir(LOCAL_STORAGE_FOLDER):
+            this_pipeline_files = [
+                file for file in os.listdir(LOCAL_STORAGE_FOLDER) if file.startswith(f"{self.pipeline_id}_")
+            ]
+            for file in this_pipeline_files:
+                os.remove(os.path.join(LOCAL_STORAGE_FOLDER, file))
```

Comment:
maybe we can use `filter` to shorten this using list comprehension

Reply:
still relevant, although in another file i tihnk

Reply:
It's already changed. The function now is 
```
        if os.path.isdir(self.trigger_sample_directory):
            this_trigger_files = list(
                filter(
                    lambda file: file.startswith(f"{pipeline_id}_{trigger_id}_"),
                    os.listdir(self.trigger_sample_directory),
                )
            )

            for file in this_trigger_files:
                os.remove(os.path.join(self.trigger_sample_directory, file))
```

Let me know if you meant something different

Reply:
I mean, the only other change I can think of is to do
```
[ os.remove(os.path.join(self.trigger_sample_directory, file)) for file in this_trigger_files]
```
 but I don't know if that's better

Reply:
should be fine the way it is